### PR TITLE
Fix files changed and reverting merge commits from the history panel bug

### PR DESCRIPTION
**Pull request by basokant**
Fix for #621 

The goals for this PR are to:

- [x] Correctly display the files changed for any merge commit
- [x] Allow for correctly reverting a merge commit
- [x] Correctly display diffs involving merge commits
- [ ] Add integration tests

### CODE REVIEW

**Comment by basokant** [comment]
Currently, the files changed in a merge commit can be displayed in the "Changed" area. However, some merge commits show no files changed (notably, merge commits created by merged pull requests).

https://user-images.githubusercontent.com/70033855/221687174-d418bf3a-73b0-478c-b7a8-28fd74cfb515.mov

File: jupyterlab_git/git.py [hunk 1]
```diff
@@ -567,12 +567,13 @@ async def log(self, path, history_count=10, follow_path=None):
 
     async def detailed_log(self, selected_hash, path):
         """
-        Execute git log -1 --numstat --oneline -z command (used to get
+        Execute git log -m -1 --numstat --oneline -z command (used to get
```
Comment (fcollonval) [inline jupyterlab_git/git.py:570]: So I guess the trick is in the various options for [`--diff-merges`](https://git-scm.com/docs/git-log#_diff_formatting), isn't it?

I'm looking forward to hearing what you learn by trying different options.

**Comment by basokant** [comment]
Add `-m --cc` to the detailed `git log`

- Detailed logs now use the options `-m` (specifies to split each merge, computing a diff for each parent of the merge)
- On it's own, this will show files changed compared to the base branch, and separately the files that were changed compared to the merged in branch. 
   - This is not ideal, as it shows too many files changed, and the virtual commits for each parent changes the way we must parse the log.
- The `--cc` produces a **combined diff**, which solves both of these problems.

https://stackoverflow.com/questions/37801342/using-git-log-to-display-files-changed-during-merge

As for reverting, we add `-m 1`, to revert to the tree of the first parent of the merge.

https://stackoverflow.com/a/7100005

**Comment by basokant** [comment]
Here is a demo of the fixes:

https://user-images.githubusercontent.com/70033855/223927554-7a3e61ec-c949-46c9-af9a-6ee3812796f0.mov

**Comment by basokant** [comment]
image

I got my environment all fixed up, and after some rewriting the integration tests for merge commits now work!

**Review by fcollonval — APPROVED** [review]
Thanks @basokant

**Pull request merged**